File: EduLink.Application/UseCases/PublicarServicioUseCase.cs
using EduLink.Application.DTOs;
using EduLink.Application.Interfaces;
using EduLink.Domain.Entities;
using EduLink.Domain.Enums;

namespace EduLink.Application.UseCases;

public class PublicarServicioUseCase
{
    private readonly IProveedorRepository _proveedorRepo;
    private readonly IServicioRepository _servicioRepo;

    public PublicarServicioUseCase(
        IProveedorRepository proveedorRepo,
        IServicioRepository servicioRepo)
    {
        _proveedorRepo = proveedorRepo;
        _servicioRepo = servicioRepo;
    }

    public async Task EjecutarAsync(PublicarServicioRequest request)
    {
        var proveedor = await _proveedorRepo.ObtenerPorIdAsync(request.ProveedorId);
        if (proveedor == null)
            throw new ArgumentException("Proveedor no encontrado.");

        var nuevoServicio = new Servicio
        {
            Titulo = request.Titulo,
            Descripcion = request.Descripcion,
            PrecioBase = request.PrecioBase,
            DuracionMinutos = request.DuracionMinutos,
            Modalidad = request.Modalidad,
            Ubicacion = request.Ubicacion
        };

        proveedor.PublicarServicio(nuevoServicio);

        await _servicioRepo.GuardarAsync(nuevoServicio);
    }
}

File: app/services/publicar_servicio_use_case.rb
class PublicarServicioUseCase
  def initialize(proveedor_repo:, servicio_repo:)
    @proveedor_repo = proveedor_repo
    @servicio_repo = servicio_repo
  end

  def ejecutar(proveedor_id:, titulo:, descripcion:, precio_base:, duracion_minutos:, modalidad:, ubicacion: nil, politica_cancelacion:)
    proveedor = @proveedor_repo.find(proveedor_id)
    raise ArgumentError, 'Proveedor no encontrado.' unless proveedor.is_a?(Entities::Proveedor)


    provider_services_count = Service.where(user_id: proveedor.id).count
    if provider_services_count >= AppConfig.limite_servicios_gratuitos
      # L+imite de servicios publicados alcanzado establecido por la tablita de parametrización
      raise StandardError, "Has alcanzado el límite de #{AppConfig.limite_servicios_gratuitos} servicios publicados."
    end

    nuevo_servicio = Entities::Servicio.new(
      titulo: titulo,
      descripcion: descripcion,
      precio_base: precio_base,
      duracion_minutos: duracion_minutos,
      modalidad: modalidad,
      ubicacion: ubicacion
    )

    nuevo_servicio.politica_cancelacion = Entities::PoliticaCancelacion.new(politica_cancelacion)

    proveedor.publicar_servicio(nuevo_servicio)

    @servicio_repo.save(nuevo_servicio)
  end
end
